src/service/ratelimit.go

```
@@ -239,10 +243,67 @@ func (this *service) shouldRateLimitWorker(
 		this.stats.GlobalShadowMode.Inc()
 	}
 
+	// If response dynamic data enabled, set dynamic data on response.
+	if this.responseDynamicMetadataEnabled {
+		response.DynamicMetadata = ratelimitToMetadata(request)
+	}
+
 	response.OverallCode = finalCode
 	return response
 }
 
+func ratelimitToMetadata(req *pb.RateLimitRequest) *structpb.Struct {
+	dm, _ := structpb.NewStruct(nil)
```

for NewStruct, I think you can look at getClientNode in xds_grpc_sotw_provider.go as an existing pattern in this Project to avoid silently ignored errors:

```
suggestion
func ratelimitToMetadata(req *pb.RateLimitRequest) *structpb.Struct {
    fields := make(map[string]*structpb.Value)
    
    // Domain
    fields["domain"] = structpb.NewStringValue(req.Domain)
    
    // Descriptors
    descriptorsValues := make([]*structpb.Value, 0, len(req.Descriptors))
    for _, descriptor := range req.Descriptors {
        s := descriptorToStruct(descriptor)
        if s == nil {
            continue
        }
        descriptorsValues = append(descriptorsValues, structpb.NewStructValue(s))
    }
    fields["descriptors"] = structpb.NewListValue(&structpb.ListValue{
        Values: descriptorsValues,
    })
    
    // HitsAddend
    if hitsAddend := req.GetHitsAddend(); hitsAddend != 0 {
        fields["hitsAddend"] = structpb.NewNumberValue(float64(hitsAddend))
    }
    
    return &structpb.Struct{Fields: fields}
}
```